frontend/src/metabase/query_builder/components/filters/modals/InlineCategoryPicker/InlineCategoryPicker.tsx

```diff
@@ -0,0 +1,162 @@
+
+import Warnings from "metabase/query_builder/components/Warnings";
+import Checkbox from "metabase/core/components/CheckBox";
+
+import { MAX_INLINE_CATEGORIES } from "./constants";
+import {
+  PickerContainer,
+  PickerGrid,
+  Loading,
+} from "./InlineCategoryPicker.styled";
+import { BulkFilterSelect } from "../BulkFilterSelect";
+
+const mapStateToProps = (state: any, props: any) => {
+  const fieldId = props.dimension?.field?.()?.id;
+  const fieldValues =
+    fieldId != null
+      ? Fields.selectors.getFieldValues(state, {
+          entityId: fieldId,
+        })
+      : [];
+  return {
+    fieldValues: fieldValues || [],
+  };
+};
+
+const mapDispatchToProps = {
+  fetchFieldValues: Fields.actions.fetchFieldValues,
+};
+
+interface InlineCategoryPickerProps {
+  query: StructuredQuery;
+  filter?: Filter;
+  newFilter: Filter;
+  dimension: Dimension;
+  fieldValues: any[];
+  fetchFieldValues: ({ id }: { id: number }) => Promise<any>;
+  handleChange: (newFilter: Filter) => void;
+  handleClear: () => void;
+}
+
+export function InlineCategoryPickerComponent({
+  query,
+  filter,
+  newFilter,
+  handleChange,
+  fieldValues,
+  fetchFieldValues,
+  dimension,
+  handleClear,
+}: InlineCategoryPickerProps) {
+  const safeFetchFieldValues = useSafeAsyncFunction(fetchFieldValues);
+
+  const shouldFetchFieldValues = !dimension?.field()?.hasFieldValues();
+  const [isLoading, setIsLoading] = useState(shouldFetchFieldValues);
+  const [hasError, setHasError] = useState(false);
+
+  useEffect(() => {
+    if (!shouldFetchFieldValues) {
+      setIsLoading(false);
+      return;
+    }
+    const field = dimension.field();
+    safeFetchFieldValues({ id: field.id })
+      .then(() => {
+        setIsLoading(false);
+      })
+      .catch(() => {
+        setHasError(true);
+      });
+  }, [dimension, safeFetchFieldValues, shouldFetchFieldValues]);
+
+  if (hasError) {
+    return (
+      <Warnings
+        warnings={[
+          t`There was an error loading the field values for this field`,
+        ]}
+      />
+    );
+  }
+
+  if (isLoading) {
+    return <Loading size={20} />;
+  }
+
+  if (fieldValues.length <= MAX_INLINE_CATEGORIES) {
+    return (
+      <SimpleCategoryFilterPicker
+        filter={filter ?? newFilter}
+        handleChange={handleChange}
+        options={fieldValues.flat()}
+      />
+    );
+  }
+
+  return (
+    <BulkFilterSelect
+      query={query}
+      filter={filter}
+      dimension={dimension}
+      handleChange={handleChange}
+      handleClear={handleClear}
+    />
+  );
+}
+
+interface SimpleCategoryFilterPickerProps {
+  filter: Filter;
+  options: (string | number)[];
+  handleChange: (newFilter: Filter) => void;
+}
+
+export function SimpleCategoryFilterPicker({
+  filter,
+  options,
+  handleChange,
+}: SimpleCategoryFilterPickerProps) {
+  const filterValues = useMemo(() => filter.arguments().filter(Boolean), [
+    filter,
+  ]);
+
+  const onChange = useCallback(
```

Comment:
This is debatable, but I'd refrain from slapping `useCallback` everywhere unless it's needed e.g. the callback would be a part of other hooks' dependencies.

This is a great article from Kent if you haven't seen https://kentcdodds.com/blog/usememo-and-usecallback

@ranquild Any thought on this?

Reply:
I see that this callback is used in a way that prevents any optimizations of `useCallback` from working https://github.com/metabase/metabase/pull/23341/files/7a3f44e8cd6e66b6c66ff672c352fc7e4150eba1#diff-ce9a7c413a40bc4c5cfb0f015f26025623e99fd72a3fc382651d6555098bba48R149. @WiNloSt is right - it doesn't work here, so maybe we should remove `useCallback` or add a nested component with `useCallback` to avoid the issue with passing additional `option` parameter. I'm fine with leaving it as it is fwiw.

nit: we use `handleXXX` naming and not `onXXX` for callbacks. We use `onXXX` only for props.